pandas/core/internals.py

```diff
@@ -5182,7 +5182,15 @@ def concatenate_block_managers(mgrs_indexers, axes, concat_axis, copy):
 
     for placement, join_units in concat_plan:
 
-        if is_uniform_join_units(join_units):
+        if len(join_units) == 1 and not join_units[0].indexers:
+            b = join_units[0].block
+            values = b.values
+            if copy:  # and values.base is not None:
+                values = values.copy()
+            elif not copy:
```

Comment:
is there a reason for this special case? IOW, why can't you handle this in ``concat_same_type`` (maybe pass in ``copy`` flag if that helps), IOW you can test there that you only have 1 block and write the same code.

Reply:
I added the separate special case because it occurs in different cases (and so how it is here, is more close to the original code), but could add it to the other special case. But, that makes `concat_same_type` more complex. Currently it does not need to bother about copying the data or not in this special case of `to_concat` values of 1, because concatenate always takes a copy.

Reply:
ok I guess. Still like to come back in and re-org this to make as clear as possible during next release. Do appreciate the support for CustomBlocks, though worry we are adding *even* more technical debt here.

Reply:
yeah what I would actually do is define a ``_concatentor`` for the base case (e.g. where we have ``staticmethod(np.concatenate)``; in ``pandas/core/dtypes/concat.py``, call it ``_concat_arrays`` or something. This (and you would have to modify all the others to take a ``copy=`` kwarg, then this code become much simpler & less special cased.

Then *all* of the concat logic is in a single place and copies are handled there. Pls create an issue for 0.22 for this (you may already have one).